File: AUTOMATION/deploy.js
import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// === CONFIGURATION ===
const config = {
  host: '192.168.2.1',
  user: 'root',
  remotePath: '/www/cgi-bin/kendalinet',
  localPath: path.join(__dirname, 'kendalinet')
};
// =====================

function deploy() {
  console.log(`🚀 Starting Push to ${config.user}@${config.host}...`);

  try {
    // 1. Ensure remote directory exists
    console.log('📁 Ensuring remote directory exists...');
    execSync(`ssh ${config.user}@${config.host} "mkdir -p ${config.remotePath}"`, { stdio: 'inherit' });

    // 2. Get list of files
    const files = fs.readdirSync(config.localPath).filter(f => f.endsWith('.sh'));

    if (files.length === 0) {
      console.error('❌ No .sh scripts found!');
      return;
    }

    // 3. Push all .sh files using a single SCP command to reduce password prompts
    console.log(`📤 Pushing all .sh scripts...`);
    try {
      // Use -O for legacy protocol (Dropbear compatibility)
      // On Windows, the wildcard expansion might be tricky, so we specify the folder
      execSync(`scp -O "${config.localPath}\\*.sh" ${config.user}@${config.host}:${config.remotePath}/`, { stdio: 'inherit' });
      console.log(`  ✅ All scripts uploaded`);
    } catch (e) {
      // If wildcard fails, fallback to individual files but warn about passwords
      console.warn('⚠️ Wildcard push failed, trying individual files...');
      for (const file of files) {
        const localFilePath = path.join(config.localPath, file);
        execSync(`scp -O "${localFilePath}" ${config.user}@${config.host}:${config.remotePath}/`, { stdio: 'inherit' });
        console.log(`  ✅ ${file} uploaded`);
      }
    }

    // 4. Set permissions and restart service
    console.log('🔑 Setting permissions and restarting web server...');
    const remoteCmd = [
      `chmod 755 ${config.remotePath}`,
      `chmod +x ${config.remotePath}/*.sh`,
      `chown root:root ${config.remotePath}/*.sh`,
      '/etc/init.d/uhttpd restart'
    ].join(' && ');

    execSync(`ssh ${config.user}@${config.host} "${remoteCmd}"`, { stdio: 'inherit' });

    console.log('\n✨ Deployment Complete!');
    console.log(`🔗 Test Link: http://${config.host}/cgi-bin/kendalinet/status.sh`);

  } catch (error) {
    console.error('\n❌ Deployment failed:');
    console.error(error.message);
  }
}

deploy();

File: AUTOMATION/deploy.py
import os
import subprocess
import sys

# === CONFIGURATION ===
ROUTER_IP = "192.168.2.1"
ROUTER_USER = "root"
REMOTE_CGI_DIR = "/www/cgi-bin/kendalinet"
LOCAL_SCRIPTS_DIR = os.path.join(os.path.dirname(__file__), "kendalinet")
# =====================

def run_command(command):
    try:
        print(f"Executing: {command}")
        result = subprocess.run(command, shell=True, check=True, capture_output=True, text=True)
        return result.stdout
    except subprocess.CalledProcessError as e:
        print(f"Error: {e.stderr}")
        return None

def main():
    print(f"🚀 Starting Push to {ROUTER_USER}@{ROUTER_IP}...")

    # 1. Create remote directory if not exists
    print("📁 Ensuring remote directory exists...")
    ssh_cmd = f"ssh {ROUTER_USER}@{ROUTER_IP} \"mkdir -p {REMOTE_CGI_DIR}\""
    run_command(ssh_cmd)

    # 2. Get list of files to push
    files = [f for f in os.listdir(LOCAL_SCRIPTS_DIR) if f.endswith('.sh')]
    
    if not files:
        print("❌ No .sh scripts found in LOCAL_SCRIPTS_DIR")
        return

    # 3. Push files using SCP
    print(f"📤 Pushing {len(files)} scripts...")
    for file in files:
        local_path = os.path.join(LOCAL_SCRIPTS_DIR, file)
        scp_cmd = f"scp \"{local_path}\" {ROUTER_USER}@{ROUTER_IP}:{REMOTE_CGI_DIR}/"
        if run_command(scp_cmd) is not None:
            print(f"  ✅ {file} uploaded")

    # 4. Set permissions and restart uhttpd
    print("🔑 Setting permissions and restarting web server...")
    remote_cmd = (
        f"chmod 755 {REMOTE_CGI_DIR} && "
        f"chmod +x {REMOTE_CGI_DIR}/*.sh && "
        f"chown root:root {REMOTE_CGI_DIR}/*.sh && "
        "/etc/init.d/uhttpd restart"
    )
    ssh_final_cmd = f"ssh {ROUTER_USER}@{ROUTER_IP} \"{remote_cmd}\""
    
    if run_command(ssh_final_cmd) is not None:
        print("\n✨ Deployment Complete!")
        print(f"🔗 Test Link: http://{ROUTER_IP}/cgi-bin/kendalinet/status.sh")
    else:
        print("\n❌ Deployment failed at the final step.")

if __name__ == "__main__":
    main()
